parcels/particlefile.py

```diff
@@ -24,46 +34,60 @@ def _is_particle_started_yet(particle, time):
 
 
 class ParticleFile(object):
-    """Initialise netCDF4.Dataset for trajectory output.
-
-    The output follows the format outlined in the Discrete Sampling Geometries
-    section of the CF-conventions:
-    http://cfconventions.org/cf-conventions/v1.6.0/cf-conventions.html#discrete-sampling-geometries
-
-    The current implementation is based on the NCEI template:
-    http://www.nodc.noaa.gov/data/formats/netcdf/v2.0/trajectoryIncomplete.cdl
-
-    Developer note: We cannot use xray.Dataset here, since it does not yet allow
-    incremental writes to disk: https://github.com/pydata/xarray/issues/199
-
+    """Initialise trajectory output.
     :param name: Basename of the output file
     :param particleset: ParticleSet to output
     :param outputdt: Interval which dictates the update frequency of file output
                      while ParticleFile is given as an argument of ParticleSet.execute()
                      It is either a timedelta object or a positive double.
     :param write_ondelete: Boolean to write particle data only when they are deleted. Default is False
-    :param chunksizes: 2d vector of sizes for NetCDF chunking. Lower values means smaller files, but also slower IO.
-                       See e.g. https://www.unidata.ucar.edu/blogs/developer/entry/chunking_data_choosing_shapes
     """
 
-    def __init__(self, name, particleset, outputdt=np.infty, write_ondelete=False, chunksizes=None):
+    def __init__(self, name, particleset, outputdt=np.infty, write_ondelete=False):
 
         self.name = name
-        self.chunksizes = [max([len(particleset), 1]), 1] if chunksizes is None else chunksizes
         self.write_ondelete = write_ondelete
         self.outputdt = outputdt
-        self.lasttraj = 0  # id of last particle written
         self.lasttime_written = None  # variable to check if time has been written already
 
         self.dataset = None
+        self.metadata = {}
         self.particleset = particleset
-
-    def open_dataset(self):
-        extension = path.splitext(str(self.name))[1]
+        self.var_names = []
+        self.var_names_once = []
+        for v in self.particleset.ptype.variables:
+            if v.to_write == 'once':
+                self.var_names_once += [v.name]
+            elif v.to_write is True:
+                self.var_names += [v.name]
+        if len(self.var_names_once) > 0:
+            self.written_once = []
+            self.file_list_once = []
+
+        self.file_list = []
+        self.time_written = []
+        self.maxid_written = -1
+        self.dataset_open = True
+
+        self.npy_path = os.path.join(os.path.dirname(str(self.name)), "out-%s" % ''.join(random.choice(string.ascii_uppercase) for _ in range(8)))
```

Comment:
It would be great for the user to be able to change this path.  This would be useful for, e.g., having a local SSD for NPY files and only build the final netCDF in distributed storage.  (We, e.g., ran into an inode-quota problem when writing the many temporary NPY files to where the final netCDF file is meant to be.)

Ways to implement this would be:

- Use env vars like `$TMP`.  But not sure if this is a good choice, because the NPY files can be very big.
- Add kwarg to `ParticleFile` and have some good default.

I'd prefer the latter.

Reply:
cc: @schmidt-christina

Reply:
Thanks @willirath! I've now implemented the kwarg version (via a new optional argument `tempwritedir` in the `ParticleFile` constructor) in a89bdaf